Revision to package access proposal (#1992)
apple/swift-evolution: proposals/0386-package-access-modifier.md

### Package Names

> Two modules belong to the same package if they were built with the same package name.  A package name must be a unique string which starts with a letter or `_`, followed by characters allowed in [this c99 extended identifier  definition](https://github.com/apple/swift-tools-support-core/blob/435a2708a6e486d69ea7d7aaa3f4ad243bc3b408/Sources/TSCUtility/StringMangling.swift#L43), including alphanumeric characters, a hypen, and a dot; any character not allowed in the definition is transcoded to a `_`.  The string is case-insensitive, and it is passed to the Swift frontend via a new flag `-package-name`.  

Reviewer: Maybe we should just say that (1) the frontend expects a UTF-8 string and will honor the uniqueness of whatever string you give it and (2) SPM does not guarantee what string it will use except that it will make a best effort to use different strings for different packages.  Case insensitivity and all this other stuff feels like unnecessary detail.

### Package Names

> Two modules belong to the same package if they were built with the same package name.  A package name must be (1) a unique identity string and (2) a valid UTF-8 string starting with a letter or `_`; for example, a string such as "Προϊόν" (translation "product") must be converted to "Προι_ο_ν" to be processed. A common package name includes alphanumeric characters (starting with a letter), `-`, `_`, and `.`.

Reviewer 2: After the rewrite, I'm not sure that the sentence `a string such as "Προϊόν" (translation "product") must be converted to "Προι_ο_ν"` naturally follows from the preceding two requirements anymore; the requirements just state that it is unique and UTF-8.

Author: updated

> Two modules belong to the same package if they were built with the same package name.  A package name must be (1) a unique identity string and (2) a valid UTF-8 string starting with a letter or `_`. A commonly accepted name includes alphanumeric characters (starting with a letter), `-`, `_`, and `.`. A build system may define its own rule to validate the input string; for example, SwiftPM checks the input against a rule defined [here](https://github.com/apple/swift-tools-support-core/blob/435a2708a6e486d69ea7d7aaa3f4ad243bc3b408/Sources/TSCUtility/StringMangling.swift#L43) and transcodes any undefined character to a `_`.  

Reviewer: Do we need *any* restrictions on this?  How about this?

```suggestion
Swift as a language leaves it up to the build system to define the boundaries of a package.  The compiler considers two modules to belong to the same package if they were built with the same package name, which is just a Unicode string.  The package name is not exposed in the source language, so its exact contents are not significant as long as it is unique to a "package".
```

When building the `Engine` module, the package name `gamePkg` is recorded in the built interface to the module.  When building `Game`, its package name `gamePkg` is compared with the package name recorded in `Engine`'s built interface; since they match, `Game` is allowed to access `Engine`'s `package` declarations.  When building `App`, its package name `appPkg` is different from `gamePkg`, so it is not allowed to access `package` symbols in either `Engine` or `Game`, which is what we want.

> If `-package-name` is not given, the `package` access modifier is disallowed.  Swift code that does not use `package` access will continue to build without needing to pass in `-package-name`.

Reviewer: ```suggestion
If `-package-name` is not given, the `package` access modifier is disallowed.  Swift code that does not use `package` access will continue to build without needing to pass in `-package-name`.  Modules built without a package name are never considered to be in the same package as any other module.
```

> The Swift Package Manager already has a concept of a package identity string for every package.  This string is verified to be unique, and it already works as a package name, so SwiftPM will pass it down automatically.  Other build systems such as Bazel may need to introduce a new build setting for a package name.  Since it needs to be unique, a reverse-DNS name may be used to avoid clashing.

Reviewer: ```suggestion
The build system should make a best effort to ensure that package names are unique.  The Swift Package Manager already has a concept of a package identity string for every package.  This string is verified to be unique, and it already works as a package name, so SwiftPM will pass it down automatically.  Other build systems such as Bazel may need to introduce a new build setting for a package name.  Since it needs to be unique, a reverse-DNS name may be used to avoid clashing.
```

### Package Symbols and `@inlinable` Functions
The Swift Package Manager already has a concept of a package identity string for every package.  This string is verified to be unique, and it already works as a package name, so SwiftPM will pass it down automatically.  Other build systems such as Bazel may need to introduce a new build setting for a package name.  Since it needs to be unique, a reverse-DNS name may be used to avoid clashing.

`package` functions can be made `@inlinable`.  Just like with `@inlinable public`, not all symbols are usable within the `@inlinable package` function: they must be `open`, `public`, `package`, `@usableFromInline`, or `usableFromPackageInline`.

> `@usableFromPackageInline` is a new attribute which allows a symbol to be used from `@inlinable package` functions (that are defined in the same module) without having to make the symbol `package` or `public`.  It can be used anywhere that `@usableFromInline` can be used, but the attributes cannot be combined.  A `@usableFromPackageInline` symbol must be `internal`. For example: 

Reviewer: Please add a discussion of `@usableFromPackageInline` to Alternatives Considered.  The line about what ideas to discuss in the proposal is not always clear, but certainly anything that was actively reviewed as part of a previous revision should never just be removed without comment — it was clearly worth considering when it was part of the proposal before!  So it should always get a mention in either Future Directions or Alternatives Considered.

```
enum Group {
  case package
  case excluded
  case named(String)
}
```

> The default value is `.package`, and the target is built with `-package-name [package_id]`.  If set to `.excluded`, no `-package-name` is passed when building the target, thus the target has no access to any package symbols; it essentially acts as if it's a client outside of the package. This would be useful for an example app or a black-box testing target in the same package.  If set to `.named([group_id])`, `-package-name [package_id].[group_id]` is passed, and the target can only access package symbols from other targets within the same group in the package.  This essentially allows "sub-packages" within the package.

Reviewer 2: I would recommend not using `[...]` as the placeholder syntax here, because in the example `.named([group_id])`, it reads like a Swift array literal. Perhaps just using all-caps (`.named(GROUP_ID)` and `-package-name PACKAGE_ID.GROUP_ID` would be clearer?

Reviewer 3: Do we have any concrete use cases for `.named()` like we do for `.excluded`?

I was trying to find some on the review thread, but all I could come up with is many mentions of a package that consists of multiple independent set of targets in a single package (some called them nuclei), but independent sets of targets shouldn't import modules from each other so they're already not seeing `public` symbols. Are we trying to design a workaround for the issue that the build system isn't separating modules based on the dependency declarations in the package manifest?

Author: The `.named(group_id)` can provide grouping within a package, where only modules within the same subgroup can see each other's package symbols. I agree that this grouping is not related to the original problem this proposal tries to solve, but might be a nice thing to add for better organization in a package.

Reviewer 3: Right, I think I have two issues with that:
- for people who don't know which option to pick, it currently seems unclear what this option really does where the other two options are clearly defined, with example use cases
- for people who did want this option, I'm not sure if it really provides what they want, since public API remains accessible from targets outside of the group (which makes sense in the context of this proposal of course, but not so much in the larger context)

I wonder if we can just have `.package` and `.excluded` in this proposal and push `.named()` to future directions which could promise a potentially more holistic feature around organizing your packages into multiple subcomponents.

Author: I was just thinking the same thing :) -- introduce `.named()` in the future. We can see if `.excluded` is enough for v1 and address grouping later if needed.

The above applies to a test target as well.  If a target contains package APIs that need to be tested, it would be recommended to split it into a "shell" that contains public APIs and a target that contains package APIs to be tested with the parameter `accessPublicOnly` set to the default value `false`.  This naturally eliminates the need to import the module with `@testable`.

2. Sub-packages

As a package becomes larger, it is inevitable for some targets to belong to a certain group within the package.  Even though it logically makes sense to move them out into a separate package, they often remain in the same package due to the maintenance and versioning overhead.  A potential solution is to allow defining a sub-package within a package. 

For example, sub-packages could be defined directly in the dependencies of the top level package in its manifest, like so.

```
let package = Package(
    name: "gamePkg",
    products: [
        .library(name: "Game", targets: ["Game"]),
    ],
    dependencies: [

>       Package(

Author: updated

Sometimes entire modules are meant to be private to the package that provides them.  Allowing this to be expressed directly would allow these utility modules to be completely hidden outside of the package, avoiding unwanted dependencies on the existence of the module.  It would also allow the build system to automatically namespace the module within the package, reducing the need for [explicit module aliases](https://github.com/apple/swift-evolution/blob/main/proposals/0339-module-aliasing-for-disambiguation.md) when utility modules of different packages share a name (such as `Utility`) or when multiple versions of a package need to be built into the same program.

### Package Boundary Customization

> 1. Opt-out

Reviewer: The Language Workgroup had requested this opt-out capability to be included in the proposal, not just discussed in Future Directions.  If you don't think it's a good idea to pursue that in the short term, please let me know why, and I can ask the Workgroup what we think about going forward with the proposal as-is.  But there was a lot of strong feedback in the review that this was important.

The top-level package and its 2 sub-packages would then be updated as follows.

Top-level package in directory `gamePkg`:

```
let package = Package(
    name: "gamePkg",
    products: [
        .library(name: "Game", targets: ["Game"]),
    ],
    dependencies: [
        .package("A"),
        .package("B")
    ],
    targets: [
        .target(name: "Game",
                dependencies: [
                    .product(name: "CoreA", package: "A"),
                    .product(name: "CoreB", package: "B")
                ]),
    ],
)
```

Sub-package `A` in directory 'gamePkg/A':
```
let package = Package(
        name: "A",
        products: [
            .library(name: "CoreA", targets: ["CoreA"]),
        ],
        targets: [
            .target(name: "CoreA", dependencies: ["EngineA"]),
            .target(name: "EngineA"),
        ]
      )
```

Sub-package `B` in directory `gamePkg/B`:

```
let package = Package(
        name: "B",
        products: [
            .library(name: "CoreB", targets: ["CoreB"]),
        ],
        targets: [
            .target(name: "CoreB", dependencies: ["EngineB"]),
            .target(name: "EngineB"),
        ]
      ),

```

> When building target `CoreA` in sub-package `A`, SwiftPM will pass `gamePkg.A` to `-package-name`; similarly, target `CoreB` will be built with `-package-name gamePkg.B`. The top level package ID is required to be unique (in case of SwiftPM, uniqueness is guaranteed via a registry), and appending a sub-package ID to it will ensure the uniqueness of each sub-package; if another top level package has the same ID as one of the sub-packages in this example, such as `A`, the `-package-name` input for the top level package `A` will be differentiated from the input `gamePkg.A` for the sub-package.

Reviewer: Some reviewers were concerned about requiring manifests to be divided this way; I think the discussion here at least needs to address that argument.  Users seem to feel that splitting manifest files has substantial practical / maintenance costs.  See e.g. this comment: https://forums.swift.org/t/se-0386-package-access-modifier/62808/154.
